### PR TITLE
Replace rc-tooltip implementation with react-popper-tooltip implementation

### PR DESCRIPTION
**Pull request by levimeahan**
The existing `Tooltip` implementation has a number of well-known issues ranging from
* Need to manually implement lazy loading the rc-tooltip library in a component due to its size
* Inability to specify onMouseEnter/onMouseLeave event handlers on tooltip trigger children
* Layers upon layers of required `rc-` libs making updates and maintenance to the component extremely difficult
* Uses OOCSS for styling, making it difficult to style with our existing techniques

And by far the biggest one:
* Errors in arrow position in many, many use cases, and necessity of a DOM hack to manually reposition the arrow when tooltip is against the edge of the screen

This PR replaces that implementation with a new implementation based off of `react-popper-tooltip`, which is much smaller and handles positioning/edge of screen cases very well by default. It will also be substantially easier for us to control the tooltip's behavior, styling, etc.

## API Changes
The API is virtually identical with a few exceptions that I will be correcting in Flexport monorepo when I upgrade latitude in it to include these changes. Of the 70 Tooltip usages in Flexport monorepo, only ~3-5 used these props.
* `overlayStyle` has been replaced with `maxWidth`
  * The only usage of overlayStyle was `overlayStyle={{ maxWidth: 200 }}` which made sense to put into a prop
* `align` has been removed
  * I looked at the cases that used this and they looked perfect with the new tooltip, so it appears the usage previously was to correct for rc-tooltip's positioning errors
* `trigger ("hover" | "click")` has been added
  * This optionally could be disabled as I don't think there are any concrete use cases for it, I mainly used it for testing. Of note, `react-popper-tooltip` does seem to listen to tap events as well even when trigger is set to hover, so it should work by default on mobile devices.
* `triggerClassName` has been added to allow styling of the tooltip trigger, since that is visible content that shouldn't matter to the tooltip. We -might- have to be restrictive in the future if any styles break the tooltip functionality but so far I haven't seen any issues.

## Manual Testing Plan
I created the component in Flexport monorepo, changed all Tooltip imports from `"latitude/Tooltip"` to `"components/base_candidate/tooltip/TooltipV2"` and confirmed visual outputs. Also tested in storybook both in Flexport repo and in Latitude repo.

### Core Shipment List
**Old Tooltip**
image

**New Tooltip**
image

### Carrier Booking > Search Results > Allocation Info (AllocationInfo.jsx)
**Old Tooltip (broken due to using `placement="bottomLeft"` which actually wasn't supported)**
image

**Old Tooltip (After changing to `placement="bottom"`)**
image

**New Tooltip**
image

### HS Code Input (example of max width control)
**Old Tooltip**
image

**New Tooltip**
image

### Edge of Screen Test
**Old Tooltip**
image

**New Tooltip**
image

### CODE REVIEW

File: Tooltip.jsx [hunk 1]
```diff
@@ -3,178 +3,149 @@
...
+const styles = StyleSheet.create({
+  trigger: {
+    display: "inline-block",
+  },
+  tooltipContainer: {
+    display: "flex",
+    flexDirection: "column",
+    margin: "0.4rem",
+    transition: "opacity 0.3s",
+  },
```
Comment (FermiDirak) [inline Tooltip.jsx]: Tooltip should have a z-index of 1500 so it rides over Modals
```suggestion
    zIndex: zIndices.zIndex1500AboveModal.value
  },
```

image

File: Tooltip.jsx [hunk 2]
```diff
@@ -3,178 +3,149 @@
...
+const arrowVerticalOffset = "-14px";
+
+const styles = StyleSheet.create({
+  trigger: {
+    display: "inline-block",
+  },
+  tooltipContainer: {
+    display: "flex",
+    flexDirection: "column",
+    margin: "0.4rem",
```
Comment (notandrewkaye) [inline Tooltip.jsx]: why `rem`?

File: Tooltip.jsx [hunk 3]
```diff
@@ -3,178 +3,149 @@
...
+    transition: "opacity 0.3s",
+  },
+  tooltipContent: {
+    backgroundColor: colors.grey60,
+    boxShadow: "0 0 4px rgba(0, 0, 0, 0.17)",
+    color: colors.white, padding: "12px",
```
Comment (notandrewkaye) [inline Tooltip.jsx]: hmm, why didn't prettier wrap this?

File: Tooltip.jsx [hunk 4]
```diff
@@ -3,178 +3,149 @@
...
-  return new Promise((resolve, reject) => {
-    waitForRcTooltipResolve = resolve;
-    waitForRcTooltipReject = reject;
-  });
+import {StyleSheet, css} from "aphrodite";
+import TooltipTrigger from "react-popper-tooltip";
+
+import colors from "./colors";
+
+const VERTICAL_OFFSET = 8;
```
Comment (notandrewkaye) [inline Tooltip.jsx]: nit: only use all uppercase if the var is exported (https://github.com/airbnb/javascript#naming--uppercase)
```suggestion
const verticalOffset = 8;
```

File: Tooltip.jsx [hunk 5]
```diff
@@ -3,178 +3,149 @@
...
+
+type Props = {|
+  +placement?: "top" | "bottom" | "left" | "right",
+  +overlay: React.Node,
+  +trigger?: "hover" | "click",
+  +mouseEnterDelay?: number, // seconds
+  +mouseExitDelay?: number, // seconds
+  +maxWidth?: number,
+  +children: string | React.Node,
+  +triggerClassName?: string,
```
Comment (notandrewkaye) [inline Tooltip.jsx]: Is there anything that passing the classname here does that wrapping the trigger in a div doesn't?

Comment (levimeahan) [reply Tooltip.jsx]: Specific case I used was dealing with getting `overflow: hidden` and `text-overflow: ellipsis` to work properly.

File: Tooltip.jsx (hunk 3 above)
Comment (levimeahan) [reply Tooltip.jsx]: Ah, I don't think I've enabled prettier watcher in RubyMine for Latitude repo

File: Tooltip.jsx (hunk 2 above)
Comment (levimeahan) [reply Tooltip.jsx]: Used the tooltip lib's default styles as a base, this was part of that

**Comment by levimeahan** [comment]
> Once zindex is added to the Tooltip content, this LGTM
> I've spun up your change as a pull request in the monorepo as well https://github.flexport.io/flexport/flexport/pull/54727
> Thanks for the contribution!

Sweet, there are a few changes to Tooltip usages that need to accompany the upgrade, should we add those to that PR or a separate one?

File: Tooltip.jsx (hunk 3 above)
Comment (dmnd) [reply Tooltip.jsx]: we also need to enable prettier in CI for this repo

File: .gitignore [hunk 6]
```diff
@@ -20,6 +20,7 @@ report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json
 .tags_sorted_by_file
 
 .vscode
+.idea
```
[the remark this answers is not in the document]
Comment (dmnd) [reply .gitignore]: FYI, you can put private stuff like this in `.git/info/exclude` to have git ignore it without committing to the shared `.gitignore`. That prevents questions like... what the heck is an `.idea` file?

File: tools/placeArrow.js [hunk 7]
```diff
@@ -1,40 +0,0 @@
-/**
- * TEAM: frontend_infra
- */
-/* eslint-disable flowtype/require-valid-file-annotation */
-/* eslint-disable flowtype/require-parameter-type */
-
-import positions from "positions";
```
Comment (dmnd) [inline tools/placeArrow.js]: Please remember to remove `positions` from `package.json`

File: Tooltip.jsx [hunk 8]
```diff
@@ -3,178 +3,152 @@
  * @flow strict
  */
 
-/**
- * A wrapper around rc-tooltip that code splits off the third party code into a
- * separate bundle. rc-tooltip will be downloaded only if it is needed to render
- * a tooltip. Use this instead of rc-tooltip.
- */
```
Comment (dmnd) [inline Tooltip.jsx]: `rc-tooltip` being 20kB is not the _only_ reason we codesplit it off. 

I think a good principle to follow is to async import all dependencies that is not needed for initial page rendering. That would include this new tooltip lib.

It's less important now that it's only ~3kB or whatever, but these small files do add up to large initial bundles. It's a little bit of extra work but if we do this for all latitude components then every app benefits.

File: Tooltip.jsx [hunk 9]
```diff
@@ -3,178 +3,152 @@
...
-  return Promise.all([rcPromise, styleSheetPromise]).then(
-    ([{default: RcTooltip}: {default: RcTooltipT}]) => {
-      invariant(
-        waitForRcTooltipResolve !== null,
-        "Resolve function must be present"
-      );
-      waitForRcTooltipResolve(RcTooltip);
-    },
-    waitForRcTooltipReject
+function MakeTrigger(children: string | React.Node, className: string) {
```
Comment (dmnd) [inline Tooltip.jsx]: This isn't a React component so it shouldn't begin with a capital letter.

If you want it to be a component, call it like `<MakeTrigger className={...} />` instead. I think this is probably the right move.

```suggestion
function makeTrigger(children: string | React.Node, className: string) {
```

File: Tooltip.jsx [hunk 10]
```diff
@@ -3,178 +3,152 @@
...
 
-/**
- * Reset module state. Only useful during testing.
- */
-export function insideTestResetModuleState() {
-  RcTooltip = null;
-  waitForRcTooltipResolve = null;
-  waitForRcTooltipReject = null;
-  waitForRcTooltip = initModuleState();
+function MakeContainer(children: React.Node, maxWidth: ?number) {
```
Comment (dmnd) [inline Tooltip.jsx]: Also not a React component

```suggestion
function makeContainer(children: React.Node, maxWidth: ?number) {
```

File: Tooltip.jsx [hunk 11]
```diff
@@ -3,178 +3,152 @@
...
+  +triggerClassName?: string,
+|};
+
+export default function Tooltip({
+  placement = "right",
+  overlay,
+  trigger = "hover",
+  children,
+  mouseEnterDelay = 0,
+  mouseExitDelay = 0,
```
Comment (dmnd) [inline Tooltip.jsx]: I think previously this default was `0.1`?

File: Tooltip.jsx [hunk 12]
```diff
@@ -3,178 +3,152 @@
...
+      tooltip={MakeContainer(overlay, maxWidth)}
+      delayShow={mouseEnterDelay * 1000}
+      delayHide={mouseExitDelay * 1000}
+      modifiers={{
+        offset: {
+          offset: `0,${verticalOffset}px`,
+        },
+      }}
+    >
+      {MakeTrigger(children, triggerClassName)}
```
Comment (dmnd) [inline Tooltip.jsx]:
```suggestion
      <Trigger className={triggerClassName} children={children} />
```

File: Tooltip.jsx [hunk 13]
```diff
@@ -3,178 +3,152 @@
...
+  mouseEnterDelay = 0,
+  mouseExitDelay = 0,
+  maxWidth = null,
+  triggerClassName = "",
+}: Props) {
+  return (
+    <TooltipTrigger
+      placement={placement}
+      trigger={trigger}
+      tooltip={MakeContainer(overlay, maxWidth)}
```
Comment (dmnd) [inline Tooltip.jsx]:
```suggestion
      tooltip={<Container overlay={overlay} maxWidth={maxWidth} />}
```

File: yarn.lock [hunk 14]
```diff
@@ -10925,58 +10881,6 @@ raw-loader@^2.0.0:
...
-rc-util@^4.0.4, rc-util@^4.4.0:
-  version "4.11.0"
-  resolved "https://registry.yarnpkg.com/rc-util/-/rc-util-4.11.0.tgz#cf437dcff74ca08a8565ae14f0368acb3a650796"
-  integrity sha512-nB29kXOXsSVjBkWfH+Z1GVh6tRg7XGZtZ0Yfie+OI0stCDixGQ1cPrS6iYxlg+AV2St6COCK5MFrCmpTgghh0w==
-  dependencies:
-    add-dom-event-listener "^1.1.0"
-    babel-runtime "6.x"
-    prop-types "^15.5.10"
-    react-lifecycles-compat "^3.0.4"
-    shallowequal "^0.2.2"
```
Comment (dmnd) [inline yarn.lock]: :tada:

File: Tooltip.jsx (hunk 11 above)
Comment (levimeahan) [reply Tooltip.jsx]: Ahh good catch was actually 0.15. Will fix 👍

File: Tooltip.jsx [hunk 15]
```diff
@@ -3,178 +3,152 @@
...
+
+import colors from "./colors";
+import {zIndices} from "./tools/zIndices";
+
+const verticalOffset = 8;
+
+type Props = {|
+  +placement?: "top" | "bottom" | "left" | "right",
+  +overlay: React.Node,
+  +trigger?: "hover" | "click",
```
[the remark this answers is not in the document]
Comment (levimeahan) [reply Tooltip.jsx]: 005804c

File: tools/placeArrow.js (hunk 7 above)
Comment (levimeahan) [reply tools/placeArrow.js]: 2eef42f

File: flow-typed/npm/react-popper-tooltip_vx.x.x.js [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+ * This is an autogenerated libdef stub for:
+ *
+ *   'react-popper-tooltip'
+ *
+ * Fill this stub out by replacing all the `any` types.
+ *
+ * Once filled out, we encourage you to share your work with the
+ * community by sending a pull request to:
+ * https://github.com/flowtype/flow-typed
+ */
```
[the remark this answers is not in the document]
Comment (levimeahan) [reply flow-typed/npm/react-popper-tooltip_vx.x.x.js]: d06261e moved to `flow-typed-overrides`, will upstream in separate PR

File: Tooltip.jsx (hunk 11 above)
Comment (levimeahan) [reply Tooltip.jsx]: c1e7a3b

File: Tooltip.jsx [hunk 17]
```diff
@@ -3,178 +3,152 @@
...
+  +mouseEnterDelay?: number, // seconds
+  +mouseExitDelay?: number, // seconds
+  +maxWidth?: number,
+  +children: string | React.Node,
```
[the remark this answers is not in the document]
Comment (levimeahan) [reply Tooltip.jsx]: c1e7a3b

File: Tooltip.jsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (levimeahan) [reply Tooltip.jsx]: Done

File: Tooltip.jsx (hunk 13 above)
Comment (levimeahan) [reply Tooltip.jsx]: Changed from `MakeContainer` to `makeContainer`, simplest way to make this work with flow and whatnot

File: Tooltip.jsx (hunk 12 above)
Comment (levimeahan) [reply Tooltip.jsx]: Changed from `MakeTrigger` to `makeTrigger`, simplest way to make this work with flow and whatnot

File: Tooltip.jsx (hunk 10 above)
Comment (levimeahan) [reply Tooltip.jsx]: Fixed

File: Tooltip.jsx (hunk 8 above)
Comment (levimeahan) [reply Tooltip.jsx]: Makes sense, will look into this in future PR

Comment (dmnd) [reply Tooltip.jsx]: That's fine 😄

**Pull request merged**